datastore/importers/nmdb-import-nmap/NmapXmlParser.cpp

```diff
@@ -0,0 +1,365 @@
+}
+
+nmdo::IpAddress
+NmapXmlParser::extractHostIpAddr(const pugi::xml_node& nodeHost) const
+{
+  pugi::xml_node const nodeIpAddr = nodeHost
+    .select_node("address[@addrtype='ipv4' or @addrtype='ipv6']")
+    .node();
+
+  nmdo::IpAddress ipAddr;
+
+  std::string ipString = nodeIpAddr.attribute("addr").as_string();
+  if (!ipString.empty()) {
+    ipAddr.setAddress(ipString);
+  }
+
+  return ipAddr;
+}
+
+
+// =========================================================================
+// XML Parsing Functions
+// =========================================================================
+void
+NmapXmlParser::extractMacAndIpAddrs(const pugi::xml_node& nmapNode, Data& data)
+{ // This code block ensures that all of the ip_addrs and mac_addrs
+  // from the scan are present for the other table's foreign keys.
+  for (const auto& xHost : nmapNode.select_nodes("host")) {
+    pugi::xml_node nodeHost = xHost.node();
+
+    bool const isResponding = extractHostIsResponding(nodeHost);
+
+    nmdo::MacAddress macAddr = extractHostMacAddr(nodeHost);
+    macAddr.setResponding(isResponding);
+    auto ipAddr {extractHostIpAddr(nodeHost)};
+    macAddr.addIp(ipAddr);
+
+    data.macAddrs.push_back(macAddr);
+  }
+}
+
+void
+NmapXmlParser::extractHostnames(const pugi::xml_node& nmapNode, Data& data)
+{
+  for (const auto& xHostname :
+        nmapNode.select_nodes("host/hostnames/hostname")) {
+    pugi::xml_node nodeHostname = xHostname.node();
+    pugi::xml_node nodeHost = nodeHostname.parent().parent();
+
+    nmdo::IpAddress ipAddr = extractHostIpAddr(nodeHost);
+    
+    std::string reason {"nmap "};
+    reason.append(nodeHostname.attribute("type").as_string());
+    ipAddr.addAlias(nodeHostname.attribute("name").as_string(), reason);
+
+    data.ipAddrs.push_back(ipAddr);
+  }
+
+  for (const auto& xScript :
+        nmapNode.select_nodes("host/hostscript/script")) {
+    pugi::xml_node nodeScript = xScript.node();
+    pugi::xml_node nodeHost   = nodeScript.parent().parent();
+
+    nmdo::IpAddress ipAddr = extractHostIpAddr(nodeHost);
+
+    std::string idValue {nodeScript.attribute("id").value()};
+    if (std::string("nbstat") == idValue) {
+      std::string reason  {"nmap nbstat"};
+      std::string line    {nodeScript.attribute("output").value()};
+      std::regex  regex   {"(NetBIOS name): ([^,]+),"};
+      std::smatch match;
+      while (std::regex_search(line, match, regex)) {
+        ipAddr.addAlias(std::string(match[2]), reason);
+        line = match.suffix();
+      }
+    } else if (std::string("smb-os-discovery") == idValue) {
+      std::string reason  {"nmap smb-os-discovery"};
+      std::string line    {nodeScript.attribute("output").value()};
+      std::regex  regex   {"(Computer name|FQDN): (.+?)\n"};
+      std::smatch match;
+      while (std::regex_search(line, match, regex)) {
+        ipAddr.addAlias(std::string(match[2]), reason);
+        line = match.suffix();
+      }
+    }
```

Comment:
```suggestion
    std::string idValue {nodeScript.attribute("id").value()};
    std::string reason  {"nmap " + idValue};
    std::string line    {nodeScript.attribute("output").value()};
    std::regex  regex;
    std::smatch match;

    if (std::string("nbstat") == idValue) {
      regex = "(NetBIOS name): ([^,]+),";
    } else if (std::string("smb-os-discovery") == idValue) {
      regex = "(Computer name|FQDN): (.+?)\n";
    }

    while (std::regex_search(line, match, regex)) {
      ipAddr.addAlias(std::string(match[2]), reason);
      line = match.suffix();
    }
```

Reply:
Ya...it didn't start as a copy/paste.  It actually looked fairly different but I start abstracting things and got them to a "same" point.  I didn't do something closer to your changes after that as I was unsure if we wanted to try to just extract the `element` nodes (for `smb-os-discovery`) instead of doing regex on the `output` string.  The `nbstat` doesn't have those nice nodes, so I was debating one way or the other.  Since the code is already there, let's go with what you got.